src/main/java/org/jenkinsci/bytecode/Transformer.java

```diff
@@ -66,49 +78,83 @@ public synchronized void loadRules(Collection<? extends ClassLoader> loaders) th
+            public MethodVisitor visitMethod(int access, final String methodName, final String methodDescriptor, final String methodSignature, String[] exceptions) {
                 final MethodVisitor base = super.visitMethod(access, methodName, methodDescriptor, methodSignature, exceptions);
 
                 return new MethodVisitor(ASM5,base) {
                     @Override
                     public void visitMethodInsn(int opcode, String owner, String name, String desc, boolean itf) {
-                        modified[0] |= spec.methods.rewrite(context,opcode,owner,name,desc, itf, base);
+                        boolean _modified = spec.methods.rewrite(context,opcode,owner,name,desc, itf, base);
+                        modified[0] |= _modified;
+                        LOGGER.log(Level.FINEST, "{0}.{1}({2}) {3} modified",
+                                   new Object[] { className, methodName, methodSignature == null ? "" : methodSignature,
+                                                 _modified ? "was" : "was not" });
                     }
 
                     @Override
                     public void visitFieldInsn(int opcode, String owner, String name, String desc) {
-                        modified[0] |= spec.fields.rewrite(context,opcode,owner,name,desc, false, base);
+                        boolean _modified = spec.fields.rewrite(context,opcode,owner,name,desc, false, base);
+                        modified[0] |= _modified;
+                        LOGGER.log(Level.FINEST, "{0}.{1}({2}) {3} modified",
+                                   new Object[] { className, methodName, methodSignature == null ? "" : methodSignature,
+                                                 _modified ? "was" : "was not" });
                     }
                 };
             }
 
             @Override
             public void visitEnd() {
+                LOGGER.log(Level.FINEST, "visitEnd(1) for {0}", className);
                 context.generateCheckerMethods(cw);
                 super.visitEnd();
+                LOGGER.log(Level.FINEST, "visitEnd(2) for {0}", className);
             }
 
-        },cr.SKIP_FRAMES);
+        },ClassReader.SKIP_FRAMES);
 
-        if (!modified[0])  return image;            // untouched
+        if (!modified[0]) {
+            LOGGER.log(Level.FINER, "class {0} was not modified.", className);
+            return image;            // untouched
+        }
+        LOGGER.log(Level.FINER, "class {0} was modified.", className);
         return cw.toByteArray();
     }
+    
+    /**
+     * Inspects a byte array representation of a class and returns the version of the bytecode.
+     * @param classData the class bytecode.
+     * @return an integer representing the major version of the bytecode.
+     */
+    private int getBytecodeVersion(byte[] classData) {
+        int version = (( classData[6] & 0xFF ) << 8 ) | (classData[7] & 0xFF);
```

Comment:
yes - its a byte so if the MSB is set this will be a negative number (0xff -> -1 -> 0xffffff when promoted to an int)